README.md
# ProtoDriver
A simple Raspberry Pi web api for driving a MAX7219 dot matrix display

This package is meant to be ran as a REST API and be called from a dedicated frontend.

ProtoDriver has only just been brought to a usable level and is in no way production ready.

## API calls

***

###/ping - GET
* A simple test call to see if the server is reachable
* Returns JSON data with the word "Pong"

###/uploadimage/< filename > - POST
* Used to pass image data in the form of bytes to the server
* filename is passed including the files extension and will be used to write the new file

###/updateimage/< filename > - PUT
* used in the same way as uploadimage but with the intent of overwriting an existing file with new data
* filename is passed including the files extension and will be used to overwrite an existing file with that name

###/deleteimage/< filename > - DELETE
* used to delete a file from the server
* filename is passed including the files extension and will be used to delete an existing file with that name

###/getimage/< filename > - GET
* used to get the contents of an image from the server
* filename is passed including the files extension and will be used to target an existing file with that name

###/getallimages - GET
* returns the contents of all images currently stored on the server paired with their filenames
* [ filename , file data]

###/displayimage/< filename > - GET
* used to display an image on the led matrix
* filename is passed including the files extension and will be used to target an existing file with that name


###/cleardisplay - GET
* clears the led matrix

CHANGES.md
#Version 0.0.2

---

Released 2021-02-05

* Added a method for returning all images currently stored on the server **GetAllImages()**
* Updated README text to reflect the new project name.
* Added function summaries into README.md

#Version 0.0.1

---

Released 2021-02-01

* First publicly visible release

test.py
import requests
from time import sleep


BASE = "http://127.0.0.1:5000/"

def run_tests():
    log = ""
    response = requests.get(BASE + "ping")
    print(response)

    response = requests.post(BASE + "uploadimage/test.txt", data="Hello World")
    print(response)

    response = requests.put(BASE + "updateimage/test.txt", data="Hello Again World")
    print(response)

    response = requests.delete(BASE + "deleteimage/test.txt")
    print(response)

    response = requests.put(BASE + "updateimage/test.txt", data="I Dont Exist")
    print(response)

    response = requests.delete(BASE + "deleteimage/test.txt")
    print(response)

    response = requests.get(BASE + "getimage/test.png")
    print(response)

    response = requests.get(BASE + "displayimage/test.png")
    print(response)

    sleep(3)

    response = requests.get(BASE + "cleardisplay")
    print(response)

    input("Press Enter to exit")


if __name__ == "__main__":
    run_tests()


tests/basecase.py
import unittest

from protodriver.app import app


class BaseCase(unittest.TestCase):

    def setUp(self):
        self.app = app.test_client()



tests/test_api.py
import unittest
import platform
from io import BytesIO
from PIL import Image
from tests.basecase import BaseCase


class MyTestCase(BaseCase):
    def test_ping(self):
        response = self.app.get('/ping')
        self.assertEqual(str, type(response.json['data']))
        self.assertEqual(200, response.status_code)

    def test_uploadimage(self):
        response = self.app.post("uploadimage/test.txt", data="Hello World")
        self.assertEqual(201, response.status_code)
        self.app.delete("deleteimage/test.txt")

    def test_updateimage(self):
        self.app.post("uploadimage/test.txt", data="Hello World")
        response = self.app.put("updateimage/test.txt", data="Hello Again World")
        self.app.delete("deleteimage/test.txt")
        self.assertEqual(200, response.status_code)

    def test_deleteimage(self):
        self.app.post("uploadimage/test.txt", data="Hello World")
        response = self.app.delete("deleteimage/test.txt")
        self.assertEqual(200, response.status_code)

    def test_getimage(self):
        temp = BytesIO()
        img = Image.new('RGBA', size=(50, 50), color=(155, 0, 0))
        img.save(temp, format='png')
        self.app.post("uploadimage/test.png", data=temp.getvalue())
        response = self.app.get("getimage/test.png")
        self.app.delete("deleteimage/test.png")
        self.assertEqual(bytes, type(response.data))
        self.assertEqual(200, response.status_code)

    def test_getallimages(self):
        temp = BytesIO()
        img = Image.new('RGBA', size=(50, 50), color=(155, 0, 0))
        img.save(temp, format='png')
        self.app.post("uploadimage/test.png", data=temp.getvalue())
        self.app.post("uploadimage/test2.png", data=temp.getvalue())
        response = self.app.get("getallimages")
        self.app.delete("deleteimage/test.png")
        self.app.delete("deleteimage/test2.png")
        self.assertEqual(str, type(response.json["data"][0][0]))
        self.assertEqual(str, type(response.json["data"][0][1]))
        self.assertEqual(str, type(response.json["data"][1][0]))
        self.assertEqual(str, type(response.json["data"][1][1]))
        self.assertEqual(200, response.status_code)


    if platform.uname().machine == "armv6l":

        def test_displayimage(self):
            temp = BytesIO()
            img = Image.new('RGB', size=(8, 8), color=(255, 255, 255))
            img.save(temp, format='png')
            self.app.post("uploadimage/test.png", data=temp.getvalue())
            response = self.app.get("displayimage/test.png")
            self.app.delete("deleteimage/test.png")
            self.assertEqual(200, response.status_code)

        def test_cleardisplay(self):
            response = self.app.get("cleardisplay")
            self.assertEqual(200, response.status_code)

    else:
        print("\nskipped PI specific test cases\n")

if __name__ == '__main__':
    unittest.main()


pdapi.py
from flask import Flask, request
from flask_restful import Resource, Api, abort, reqparse
from base64 import encodebytes
from PIL import Image

import os
import io
import platform

IS_PI = False

if platform.uname().machine == "armv6l":
    import leddriver
    IS_PI = True



app = Flask(__name__)
api = Api(app)
UPLOAD_DIRECTORY = "api_uploaded_files"
ALLOWED_EXTENSIONS = {'txt', 'pdf', 'png', 'jpg', 'jpeg', 'gif'}

if not os.path.exists(UPLOAD_DIRECTORY):
    os.makedirs(UPLOAD_DIRECTORY)


def find_file(filename):
    if "/" in filename:
        abort(400)
    if not os.path.exists(os.path.join(UPLOAD_DIRECTORY, filename)):
        abort(404)


def get_response_image(image_path):
    pil_img = Image.open(image_path, mode='r') # reads the PIL image
    byte_arr = io.BytesIO()
    pil_img.save(byte_arr, format='PNG') # convert the PIL image to byte array
    encoded_img = encodebytes(byte_arr.getvalue()).decode('ascii') # encode as base64
    return encoded_img


class Ping(Resource):
    def get(self):
        return {"data": "Pong"}


class UploadImage(Resource):
    def post(self, filename):
        if "/" in filename:
            abort(400)

        if os.path.exists(os.path.join(UPLOAD_DIRECTORY, filename)):
            abort(409)

        with open(os.path.join(UPLOAD_DIRECTORY, filename), "wb") as fp:
            fp.write(request.data)

        return "File Created", 201


class UpdateImage(Resource):
    def put(self, filename):

        find_file(filename)
        with open(os.path.join(UPLOAD_DIRECTORY, filename), "wb") as fp:
            fp.write(request.data)

        return "File Created", 200


class DeleteImage(Resource):
    def delete(self, filename):

        find_file(filename)
        os.remove(os.path.join(UPLOAD_DIRECTORY, filename))


class GetImage(Resource):
    def get(self, filename):

        find_file(filename)
        return get_response_image(os.path.join(UPLOAD_DIRECTORY, filename))


class GetAllImages(Resource):
    pass


if IS_PI:

    class DisplayImage(Resource):
        def get(self, filename):
            find_file(filename)
            leddriver.display_image(os.path.join(UPLOAD_DIRECTORY, filename))
            return "Image displayed", 200

    class ClearDisplay(Resource):
        def get(self):
            leddriver.clear_display()
            return "Display Cleared", 200


api.add_resource(Ping, "/ping")
api.add_resource(UploadImage, "/uploadimage/<filename>")
api.add_resource(UpdateImage, "/updateimage/<filename>")
api.add_resource(DeleteImage, "/deleteimage/<filename>")
api.add_resource(GetImage, "/getimage/<filename>")

if IS_PI:
    api.add_resource(DisplayImage, "/displayimage/<filename>")
    api.add_resource(ClearDisplay, "/cleardisplay")

app.run(debug=True, port=5000)


protodriverpi/app.py
from flask import Flask, request
from flask_restful import Resource, Api, abort
from base64 import encodebytes
from PIL import Image

import os
import io
import platform

IS_PI = False

if platform.uname().machine == "armv6l":
    import leddriver
    IS_PI = True



app = Flask(__name__)
api = Api(app)
UPLOAD_DIRECTORY = "api_uploaded_files"
ALLOWED_EXTENSIONS = {'txt', 'pdf', 'png', 'jpg', 'jpeg', 'gif'}

if not os.path.exists(UPLOAD_DIRECTORY):
    os.makedirs(UPLOAD_DIRECTORY)


def find_file(filename):
    if "/" in filename:
        abort(400)
    if not os.path.exists(os.path.join(UPLOAD_DIRECTORY, filename)):
        abort(404)


def get_response_image(image_path):
    pil_img = Image.open(image_path, mode='r') # reads the PIL image
    byte_arr = io.BytesIO()
    pil_img.save(byte_arr, format='PNG') # convert the PIL image to byte array
    encoded_img = encodebytes(byte_arr.getvalue()).decode('ascii') # encode as base64
    return encoded_img


class Ping(Resource):
    def get(self):
        return {"data": "Pong"}


class UploadImage(Resource):
    def post(self, filename):
        if "/" in filename:
            abort(400)

        if os.path.exists(os.path.join(UPLOAD_DIRECTORY, filename)):
            abort(409)

        with open(os.path.join(UPLOAD_DIRECTORY, filename), "wb") as fp:
            fp.write(request.data)

        return "File Created", 201


class UpdateImage(Resource):
    def put(self, filename):

        find_file(filename)
        with open(os.path.join(UPLOAD_DIRECTORY, filename), "wb") as fp:
            fp.write(request.data)

        return "File Created", 200


class DeleteImage(Resource):
    def delete(self, filename):

        find_file(filename)
        os.remove(os.path.join(UPLOAD_DIRECTORY, filename))


class GetImage(Resource):
    def get(self, filename):

        find_file(filename)
        return get_response_image(os.path.join(UPLOAD_DIRECTORY, filename))


class GetAllImages(Resource):
    pass


if IS_PI:

    class DisplayImage(Resource):
        def get(self, filename):
            find_file(filename)
            leddriver.display_image(os.path.join(UPLOAD_DIRECTORY, filename))
            return "Image displayed", 200

    class ClearDisplay(Resource):
        def get(self):
            leddriver.clear_display()
            return "Display Cleared", 200


api.add_resource(Ping, "/ping")
api.add_resource(UploadImage, "/uploadimage/<filename>")
api.add_resource(UpdateImage, "/updateimage/<filename>")
api.add_resource(DeleteImage, "/deleteimage/<filename>")
api.add_resource(GetImage, "/getimage/<filename>")

if IS_PI:
    api.add_resource(DisplayImage, "/displayimage/<filename>")
    api.add_resource(ClearDisplay, "/cleardisplay")


def run(**kwargs):
    app.run(**kwargs)


if __name__ == "__main__":
    print("Starting Api...")
    run(debug=True, port=5000)
